File: tests/swift/test/srcTest.swift
//
//  test.swift
//  test
//
//  Created by Nick Bransby-Williams on 06/02/2016.
//
//

import XCTest
import JavaScriptCore
@testable import src

class test: XCTestCase {
    
    override func setUp() {
        super.setUp()
    }
    
    override func tearDown() {
        try! this.context.eval(NSBundle(identifier: "io.xrails.src")!.pathForResource("src", ofType: "js")!) as Void
    }
    
    func testSimpleException() {
        do {
            try throwSimpleError()
            XCTFail()
        } catch let e as CustomStringConvertible {
            XCTAssertEqual("Simple error message", e.description)
        } catch {
            XCTFail()
        }
    }
    
    func testSpecialException() {
        do {
            try throwSpecialError()
            XCTFail()
        } catch let e as SpecialException {
            XCTAssertEqual("Special error message", e.message)
        } catch {
            XCTFail()
        }
    }

    static var testInterfaceMethodCalled = false
    
    class SimpleInterfaceImpl : SimpleInterface {
        func voidNoArgMethod() {
            testInterfaceMethodCalled = true
        }
    }

    func testInterface() {
        XCTAssertFalse(simpleInterfaceInstanceCalled)
        simpleInterfaceInstance.voidNoArgMethod()
        XCTAssert(simpleInterfaceInstanceCalled)
        acceptSimpleInterface(SimpleInterfaceImpl())
        XCTAssert(test.testInterfaceMethodCalled)
        test.testInterfaceMethodCalled = false
        simpleInterfaceInstance.voidNoArgMethod()
        XCTAssert(test.testInterfaceMethodCalled)
    }
    
    static var testInheritanceMethodCalled = false
    
    class SimpleObjectOverride : SimpleObject {
        override func methodToOverride() {
            testInheritanceMethodCalled = true
        }
    }
    
    class SimpleObjectOverrideCallSuper : SimpleObject {
        override func methodToOverride() {
            super.methodToOverride()
        }
    }
    
    func testInheritance() {
        var o = SimpleObject(5)
        o.callOverriddenMethod()
        XCTAssert(o.methodToOverrideCalled)
        
        o = SimpleObjectOverride(5)
        o.callOverriddenMethod()
        XCTAssert(test.testInheritanceMethodCalled)
        XCTAssertFalse(o.methodToOverrideCalled)
        
        o = SimpleObjectOverrideCallSuper(5)
        o.methodToOverride()
        XCTAssert(o.methodToOverrideCalled)
    }
    
    func testObjects() {
        SimpleObject.staticVoidNoArgMethod()
        
        XCTAssertNotNil(simpleObjectInstance)
        
        XCTAssertEqual(49.0, simpleObjectInstance.numberSingleObjectArgMethod(simpleObjectInstance))
        
        XCTAssertEqual(14.0, simpleObjectInstance.numberSingleObjectArgMethod(SimpleObject(2)))
        
        XCTAssertEqual(simpleObjectInstance, simpleObjectInstance)
        XCTAssert(simpleObjectInstance == simpleObjectInstance.upcastThisToObject())
        XCTAssert(simpleObjectInstance.upcastThisToObject() == simpleObjectInstance)
        XCTAssert(simpleObjectInstance.upcastThisToObject() == simpleObjectInstance.upcastThisToObject())
    }
    
    func testFunctions() {
        voidNoArgFunction()
        XCTAssert(voidNoArgFunctionCalled)
    
        XCTAssertEqual("stringNoArgFunctionReturnValue", stringNoArgFunction())
        
        XCTAssertEqual(25.0, numberMultipleArgFunction(5, b: 5))
        XCTAssertEqual(-3.0, numberMultipleArgFunction(-2, b: 1.5))
        
        XCTAssertEqual("stringNoArgLambdaReturnValue", stringNoArgLambda())
        XCTAssertEqual(stringNoArgLambda(), stringNoArgLambda())
        
        stringNoArgLambda = { "expectedReturnValue" }
        XCTAssertEqual("expectedReturnValue", stringNoArgLambda())
    }
    
    func testConstTypes() {
        XCTAssertFalse(booleanConst)
        XCTAssert(numberConst.isNaN)
        XCTAssertEqual(stringConst, "stringConstLiteral")
        XCTAssertEqual(numberArrayConst, [1, 2, 3])
        XCTAssertEqual(stringArrayArrayConst, [["1", "2", "3"], ["4", "5", "6"], ["7", "8", "9"]])
        XCTAssertEqual(anyConst as? String, "anyConstLiteral")
    }
    
    func testOptionalConstTypes() {
        XCTAssertNil(optionalBooleanConst)
        XCTAssertNil(optionalNumberConst)
        XCTAssertNil(optionalStringConst)
        XCTAssertNil(optionalNumberArrayConst)
        XCTAssertNil(optionalNullAnyConst)
        //XCTAssertEqual(optionalNonNullAnyConst as? String, stringConst)
    }
    
    func testVarTypes() {
        XCTAssertTrue(booleanVar)
        booleanVar = booleanConst
        XCTAssertEqual(booleanVar, booleanConst)
        
        XCTAssertEqual(numberVar, 0)
        numberVar = numberConst
        XCTAssert(numberVar.isNaN)
        
        XCTAssertEqual(stringVar, "stringVarLiteral")
        stringVar = stringConst
        XCTAssertEqual(stringVar, stringConst)
        
        XCTAssertEqual(anyVar as? String, "anyVarLiteral")
        anyVar = anyConst
        XCTAssertEqual(anyVar as? String, anyConst as? String)
        
        XCTAssertEqual(numberArrayVar, [])
        numberArrayVar = numberArrayConst
        XCTAssertEqual(numberArrayVar, numberArrayConst)

        XCTAssertEqual(stringArrayArrayVar, [])
        stringArrayArrayVar = stringArrayArrayConst
        XCTAssertEqual(stringArrayArrayVar, stringArrayArrayConst)
        
        numberArrayVar = [5]
        XCTAssertEqual(numberArrayVar, [5])
        
        stringArrayArrayVar = [["yo"]]
        XCTAssertEqual(stringArrayArrayVar, [["yo"]])
    }
    
    
    func testOptionalVarTypes() {
        XCTAssertNil(optionalBooleanVar)
        optionalBooleanVar = false
        XCTAssertEqual(optionalBooleanVar, booleanConst)
        optionalBooleanVar = nil
        XCTAssertNil(optionalBooleanVar)
        
        XCTAssertNil(optionalNumberVar)
        optionalNumberVar = 4000
        XCTAssertEqual(optionalNumberVar, 4000)
        optionalNumberVar = nil
        XCTAssertNil(optionalNumberVar)
        
        XCTAssertNil(optionalStringVar)
        optionalStringVar = "stringConstLiteral"
        XCTAssertEqual(optionalStringVar, stringConst)
        optionalStringVar = nil
        XCTAssertNil(optionalStringVar)
        
        XCTAssertNil(optionalAnyVar)
        optionalAnyVar = "anyConstLiteral"
        XCTAssertEqual(optionalAnyVar! as? String, anyConst as? String)
        optionalAnyVar = nil
        XCTAssertNil(optionalAnyVar)
    }

    func testNoErasureForBasicTypes() {
        anyVar = booleanConst
        XCTAssertEqual(anyVar as? Bool, booleanConst)
    }
    
    func testErasureForNonBasicTypes() {
        XCTAssert(anyObjectInstance == anyObjectInstance)
        anyVar = anyObjectInstance
        XCTAssert(anyVar == anyObjectInstance)
    }
    
    class UnknownNativeSideObject {
        
    }
    
    func testUnknownNativeSideObjectDeallocation() {
        var o :AnyObject? = UnknownNativeSideObject()
        weak var ref = o
        XCTAssertNotNil(ref)
        optionalAnyVar = o
        o = nil
        XCTAssertNotNil(ref)
        optionalAnyVar = nil
        JSSynchronousGarbageCollectForDebugging(this.context.ref)
        XCTAssertNil(ref)
    }

    func testUnknownJSSideObjectDeallocation() {
        var o :AnyObject? = optionalAnyObjectInstance as? AnyObject
        weak var ref = o!
        XCTAssertNotNil(ref)
        optionalAnyObjectInstance = nil
        JSSynchronousGarbageCollectForDebugging(this.context.ref)
        XCTAssertNotNil(ref)
        optionalAnyVar = o
        o = nil
        XCTAssertNil(ref)
    }
    
    func testKnownNativeSideObjectDeallocation() {
        var o :AnyObject? = SimpleObject(1)
        weak var ref = o
        XCTAssertNotNil(ref)
        optionalAnyVar = o
        o = nil
        XCTAssertNotNil(ref)
        optionalAnyVar = nil
        JSSynchronousGarbageCollectForDebugging(this.context.ref)
        XCTAssertNil(ref)
    }
    
    func testKnownJSSideObjectDeallocation() {
        var o :AnyObject? = simpleInterfaceInstance
        weak var ref = o
        XCTAssertNotNil(ref)
        optionalAnyVar = o
        o = nil
        XCTAssertNotNil(ref)
        optionalAnyVar = nil
        JSSynchronousGarbageCollectForDebugging(this.context.ref)
        XCTAssertNil(ref)
    }
    
    func testNativeSideInterfaceDeallocation() {
        var o :AnyObject? = SimpleInterfaceImpl()
        weak var ref = o
        XCTAssertNotNil(ref)
        optionalAnyVar = o
        o = nil
        XCTAssertNotNil(ref)
        optionalAnyVar = nil
        JSSynchronousGarbageCollectForDebugging(this.context.ref)
        XCTAssertNil(ref)
    }
    
    func testJSSideInterfaceDeallocation() {
        var o :AnyObject? = simpleInterfaceInstance
        weak var ref = o
        XCTAssertNotNil(ref)
        optionalAnyVar = o
        o = nil
        XCTAssertNotNil(ref)
        optionalAnyVar = nil
        JSSynchronousGarbageCollectForDebugging(this.context.ref)
        XCTAssertNil(ref)
    }
}

File: tests/java/src/test/java/io/xrails/SrcTest.java
package io.xrails;

import org.junit.Before;
import org.junit.Test;

import java.lang.ref.WeakReference;
import java.util.Arrays;
import java.util.Collections;

import static io.xrails.Src.*;
import static org.junit.Assert.*;

/**
 * Created by nbransby on 20/02/2016.
 */
public class SrcTest {

    @Before
    public void setUp() {
        JS.eval("../input/src.js");
    }

    @Test
    public void testSimpleException() throws Exception {
        try {
            throwSimpleError();
            fail();
        } catch (Exception e) {
            assertEquals("Error: Simple error message", e.getMessage());
        }
    }

    @Test
    public void testSpecialException() throws Exception {
        try {
            throwSpecialError();
            fail();
        } catch (SpecialException e) {
            assertEquals("Special error message", e.message());
        }
    }

    @Test
    public void testObjects() throws Exception {
        SimpleObject.staticVoidNoArgMethod();

        assertNotNull(simpleObjectInstance());

        assertEquals(49.0, simpleObjectInstance().numberSingleObjectArgMethod(simpleObjectInstance()));

        assertEquals(14.0, simpleObjectInstance().numberSingleObjectArgMethod(new SimpleObject(2)));

        assertEquals(simpleObjectInstance(), simpleObjectInstance());
        assertEquals(simpleObjectInstance(), simpleObjectInstance().upcastThisToObject());
        assertEquals(simpleObjectInstance().upcastThisToObject(), simpleObjectInstance());
    }

    private boolean testInheritanceMethodCalled = false;

    @Test
    public void testInheritance() throws Exception {
        SimpleObject o = new SimpleObject(5);
        o.callOverriddenMethod();
        assertTrue(o.methodToOverrideCalled());

        o = new SimpleObject(5) {
            @Override
            public void methodToOverride() {
                testInheritanceMethodCalled = true;
            }
        };
        o.callOverriddenMethod();
        assertTrue(testInheritanceMethodCalled);
        assertFalse(o.methodToOverrideCalled());

        o = new SimpleObject(5) {
            @Override
            public void methodToOverride() {
                super.methodToOverride();
            }
        };
        o.methodToOverride();
        assertTrue(o.methodToOverrideCalled());
    }

    private boolean testInterfaceMethodCalled = false;

    @Test
    public void testInterface() throws Exception {
        simpleInterfaceInstance().voidNoArgMethod();
        assertTrue(simpleInterfaceInstanceCalled());

        acceptSimpleInterface(new SimpleInterface() {
            @Override
            public void voidNoArgMethod() {
                testInterfaceMethodCalled = true;
            }
        });
        assertTrue(testInterfaceMethodCalled);
        testInterfaceMethodCalled = false;
        simpleInterfaceInstance().voidNoArgMethod();
        assertTrue(testInterfaceMethodCalled);
    }

    @Test
    public void testFunctions() throws Exception {
        voidNoArgFunction();
        assertTrue(voidNoArgFunctionCalled());

        assertEquals("stringNoArgFunctionReturnValue", stringNoArgFunction());

        assertEquals(25.0, numberMultipleArgFunction(5, 5));
        assertEquals(-3.0, numberMultipleArgFunction(-2, 1.5));

        assertEquals("stringNoArgLambdaReturnValue", stringNoArgLambda().get());
        assertEquals(stringNoArgLambda().get(), stringNoArgLambda().get());
        stringNoArgLambda(() -> "expectedReturnValue");
        assertEquals("expectedReturnValue", stringNoArgLambda().get());
    }

    @Test
    public void testConstTypes() throws Exception {
        assertFalse(booleanConst());
        assertTrue(Double.isNaN(numberConst().doubleValue()));
        assertEquals("stringConstLiteral", stringConst());
        assertEquals(Arrays.asList(1, 2, 3), numberArrayConst());
        assertEquals(Arrays.asList(Arrays.asList("1", "2", "3"), Arrays.asList("4", "5", "6"), Arrays.asList("7", "8", "9")), stringArrayArrayConst());
        assertEquals("anyConstLiteral", anyConst());
        assertEquals(Arrays.asList(1, null, 3), numberOrNullArrayConst());
    }

    @Test
    public void testOptionalConstTypes() throws Exception {
        assertFalse(optionalBooleanConst().isPresent());
        assertFalse(optionalNumberConst().isPresent());
        assertFalse(optionalStringConst().isPresent());
        assertFalse(optionalNumberArrayConst().isPresent());
        assertFalse(optionalNullAnyConst().isPresent());
        assertEquals(optionalNonNullAnyConst().get(), stringConst());
    }

    @Test
    public void testVarTypes() throws Exception {
        assertTrue(booleanVar());
        booleanVar(booleanConst());
        assertEquals(booleanConst(), booleanVar());

        assertEquals(0, numberVar());
        numberVar(numberConst());
        assertTrue(Double.isNaN(numberVar().doubleValue()));

        assertEquals("stringVarLiteral", stringVar());
        stringVar(stringConst());
        assertEquals(stringConst(), stringVar());

        assertEquals("anyVarLiteral", anyVar());
        anyVar(anyConst());
        assertEquals(anyConst(), anyVar());

        assertEquals(Collections.emptyList(), numberArrayVar());
        numberArrayVar(numberArrayConst());
        assertEquals(numberArrayConst(), numberArrayVar());

        assertEquals(Collections.emptyList(), stringArrayArrayVar());
        stringArrayArrayVar(stringArrayArrayConst());
        assertEquals(stringArrayArrayConst(), stringArrayArrayVar());

        numberArrayVar(Collections.singletonList(5));
        assertEquals(Collections.singletonList(5), numberArrayVar());

        stringArrayArrayVar(Collections.singletonList(Collections.singletonList("yo")));
        assertEquals(Collections.singletonList(Collections.singletonList("yo")), stringArrayArrayVar());
    }

    @Test
    public void testOptionalVarTypes() throws Exception {
        assertFalse(optionalBooleanVar().isPresent());
        optionalBooleanVar(false);
        assertEquals(booleanConst(), optionalBooleanVar().get());
        optionalBooleanVar(null);
        assertFalse(optionalBooleanVar().isPresent());

        assertFalse(optionalNumberVar().isPresent());
        optionalNumberVar(4000);
        assertEquals(4000, optionalNumberVar().get());
        optionalNumberVar(null);
        assertFalse(optionalNumberVar().isPresent());

        assertFalse(optionalStringVar().isPresent());
        optionalStringVar("stringConstLiteral");
        assertEquals(stringConst(), optionalStringVar().get());
        optionalStringVar(null);
        assertFalse(optionalStringVar().isPresent());

        assertFalse(optionalAnyVar().isPresent());
        optionalAnyVar("anyConstLiteral");
        assertEquals(anyConst(), optionalAnyVar().get());
        optionalAnyVar(null);
        assertFalse(optionalAnyVar().isPresent());
    }

    @Test
    public void testNoErasureForBasicTypes() throws Exception {
        anyVar(booleanConst());
        assertEquals(booleanConst(), anyVar());
        assertTrue(anyVar() == booleanConst());
    }

    @Test
    public void testErasureForNonBasicTypes() throws Exception {
        anyVar(anyObjectInstance());
        assertEquals(anyObjectInstance(), anyVar());
        assertFalse(anyVar() == anyObjectInstance());
    }

    @Test
    public void testUnknownObjectDeallocation() throws Exception {
        Object o = new Object();
        WeakReference ref = new WeakReference<>(o);
        assertNotNull(ref.get());
        optionalAnyVar(o);
        o = null;
        System.gc();
        System.runFinalization();
        assertNotNull(ref.get());
        optionalAnyVar(null);
        System.gc();
        System.runFinalization();
        assertNull(ref.get());
    }

    @Test
    public void testKnownObjectDeallocation() throws Exception {
        Object o = new SimpleObject(1);
        WeakReference ref = new WeakReference<>(o);
        assertNotNull(ref.get());
        optionalAnyVar(o);
        o = null;
        System.gc();
        System.runFinalization();
        assertNotNull(ref.get());
        optionalAnyVar(null);
        System.gc();
        System.runFinalization();
        assertNull(ref.get());
    }

    @Test
    public void testNativeSideInterfaceDeallocation() throws Exception {
        SimpleInterface o = new SimpleInterface() {
            @Override
            public void voidNoArgMethod() {
            }
        };
        WeakReference ref = new WeakReference<>(o);
        assertNotNull(ref.get());
        optionalAnyVar(o);
        o = null;
        System.gc();
        System.runFinalization();
        assertNotNull(ref.get());
        optionalAnyVar(null);
        System.gc();
        System.runFinalization();
        assertNull(ref.get());
    }

    @Test
    public void testJSSideInterfaceDeallocation() throws Exception {
        SimpleInterface so = simpleInterfaceInstance();
        WeakReference ref2 = new WeakReference<>(so);
        optionalAnyVar(so);
        so = null;
        System.gc();
        System.runFinalization();
        assertNotNull(ref2.get());
        optionalAnyVar(null);
        System.gc();
        System.runFinalization();
        assertNull(ref2.get());
    }
}